source/common/http/conn_manager_utility.cc

```diff
@@ -131,6 +132,68 @@ void ConnectionManagerUtility::mutateRequestHeaders(Http::HeaderMap& request_hea
   if (config.tracingConfig()) {
     Tracing::HttpTracerUtility::mutateHeaders(request_headers, runtime);
   }
+  mutateXfccRequestHeader(request_headers, connection, config);
+}
+
+void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_headers,
+                                                       Network::Connection& connection,
+                                                       ConnectionManagerConfig& config) {
+  // When AlwaysForwardOnly is set, always forward the XFCC header without modification.
+  if (config.forwardClientCert() == Http::ForwardClientCertType::AlwaysForwardOnly) {
+    return;
+  }
+  // When Sanitize is set, or the connection is not mutual TLS, remove the XFCC header.
+  if (config.forwardClientCert() == Http::ForwardClientCertType::Sanitize ||
+      !(connection.ssl() && connection.ssl()->peerCertificatePresented())) {
+    request_headers.removeForwardedClientCert();
+    return;
+  }
+
+  // When ForwardOnly is set, always forward the XFCC header without modification.
+  if (config.forwardClientCert() == Http::ForwardClientCertType::ForwardOnly) {
+    return;
+  }
+
+  // TODO(myidpt): Handle the special characters in By and SAN fields.
+  std::vector<std::string> client_cert_details;
```

Comment:
FYI this is going to be substantially slower than what you had before. This is related to what @lizan mentioned with doing direct append into the header value. In the interest of getting this out and not doing premature optimization, let's leave this simpler code, but please add a TODO that we should look at perf in this case. (Basically anything that is done on every request is something that we need to look at perf for).